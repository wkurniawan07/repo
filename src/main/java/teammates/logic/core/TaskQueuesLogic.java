package teammates.logic.core;

import com.google.appengine.api.taskqueue.Queue;
import com.google.appengine.api.taskqueue.QueueFactory;
import com.google.appengine.api.taskqueue.TaskOptions;

import teammates.common.util.JsonUtils;
import teammates.common.util.TaskWrapper;

/**
 * Handles operations related to task queues.
 */
public class TaskQueuesLogic {

    /**
     * Adds the given task to the specified queue.
     *
     * @param task the task object containing the details of task to be added
     */
    public void addTask(TaskWrapper task) {
        addDeferredTask(task, 0);
    }

    /**
     * Adds the given task, to be run after the specified time, to the specified queue.
     *
     * @param task the task object containing the details of task to be added
     * @param countdownTime the time delay for the task to be executed
     */
    public void addDeferredTask(TaskWrapper task, long countdownTime) {
        Queue requiredQueue = QueueFactory.getQueue(task.getQueueName());
        TaskOptions taskToBeAdded = TaskOptions.Builder.withUrl(task.getWorkerUrl());
        if (countdownTime > 0) {
            taskToBeAdded.countdownMillis(countdownTime);
        }

        // GAE's Task Queue API only allows either parameter map or body, not both
        if (task.getRequestBody() == null) {
            task.getParamMap().forEach((key, value) -> taskToBeAdded.param(key, value));
        } else {
            taskToBeAdded.payload(JsonUtils.toCompactJson(task.getRequestBody()));
        }

        requiredQueue.add(taskToBeAdded);
    }

}
