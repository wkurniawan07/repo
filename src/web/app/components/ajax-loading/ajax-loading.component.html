<div class="loading-container">
  <div class="spinner-border spinner-border-sm text-{{useBlueSpinner ? 'primary' : 'white'}}" role="status"></div>
</div>
