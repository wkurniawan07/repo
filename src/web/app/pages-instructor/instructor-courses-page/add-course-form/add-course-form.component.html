<div class="card border-primary fill-plain">
  <div class="close-header">
    <button type="button" class="close" aria-label="Close" (click)="closeEditFormHandler()">
      <span aria-hidden="true">&times;</span>
    </button>
  </div>
  <div class="card-body text-center">
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label">Course ID:</label>
      <div class="col-md-6 text-md-left">
        <input id="new-course-id" class="form-control" type="text" [(ngModel)]="newCourseId"
            [maxlength]="40" [disabled]="isAddingCourse" placeholder="e.g. CS3215-2013Semester1"/>
      </div>
    </div>
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label">Course Name:</label>
      <div class="col-md-6 text-md-left">
        <input id="new-course-name" class="form-control" type="text" [(ngModel)]="newCourseName"
            [maxlength]="64" [disabled]="isAddingCourse" placeholder="e.g. Software Engineering"/>
      </div>
    </div>
    <div class="form-group row">
      <label class="col-md-3 text-md-right col-form-label ngb-tooltip-class"><span ngbTooltip="You should not
                need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                to match the current time offset in your area, including clock changes due to daylight saving time.">
        Time Zone:</span></label>
      <div class="col-md-6 text-md-left">
        <div class="input-group">
          <select id="new-time-zone" class="form-control" [(ngModel)]="timezone" [disabled]="isAddingCourse">
            <option *ngFor="let timezone of timezones" [value]="timezone.id">{{ timezone.id }} ({{ timezone.offset }})</option>
          </select>
          <span class="input-group-btn" style="margin-left: 5px;">
            <button class="btn btn-primary btn-margin-left" (click)="onAutoDetectTimezone()" [disabled]="isAddingCourse">Auto Detect</button>
          </span>
        </div>
      </div>
    </div>
    <div class="form-group row">
      <div class="col-md-9 offset-md-3 text-md-left">
        <button id="btn-save-course" class="btn btn-success" (click)="onSubmit()" [disabled]="isAddingCourse">
          <tm-ajax-loading *ngIf="isAddingCourse"></tm-ajax-loading>
          Add Course
        </button>
      </div>
    </div>
  </div>
</div>
