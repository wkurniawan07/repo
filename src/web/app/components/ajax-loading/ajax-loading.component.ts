import { Component, Input, OnInit } from '@angular/core';

/**
 * Displaying the ajax loader.
 */
@Component({
  selector: 'tm-ajax-loading',
  templateUrl: './ajax-loading.component.html',
  styleUrls: ['./ajax-loading.component.scss'],
})
export class AjaxLoadingComponent implements OnInit {

  @Input()
  useBlueSpinner: boolean = false;

  constructor() { }

  ngOnInit(): void {
  }

}
