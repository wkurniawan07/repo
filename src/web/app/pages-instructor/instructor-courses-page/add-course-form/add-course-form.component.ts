import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { finalize } from 'rxjs/operators';
import { CourseService } from '../../../../services/course.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { ErrorMessageOutput } from '../../../error-message-output';

interface Timezone {
  id: string;
  offset: string;
}

const formatTwoDigits: Function = (n: number): string => {
  if (n < 10) {
    return `0${n}`;
  }
  return String(n);
};

/**
 * Instructor add new course form
 */
@Component({
  selector: 'tm-add-course-form',
  templateUrl: './add-course-form.component.html',
  styleUrls: ['./add-course-form.component.scss'],
})
export class AddCourseFormComponent implements OnInit {

  @Input() isEnabled: boolean = true;
  @Output() courseAdded: EventEmitter<void> = new EventEmitter<void>();
  @Output() closeCourseFormEvent: EventEmitter<void> = new EventEmitter<void>();

  timezones: Timezone[] = [];
  timezone: string = '';
  newCourseId: string = '';
  newCourseName: string = '';
  isAddingCourse: boolean = false;

  constructor(private statusMessageService: StatusMessageService,
              private courseService: CourseService,
              private timezoneService: TimezoneService) { }

  ngOnInit(): void {
    for (const [id, offset] of Object.entries(this.timezoneService.getTzOffsets())) {
      const hourOffset: number = Math.floor(Math.abs(offset) / 60);
      const minOffset: number = Math.abs(offset) % 60;
      const sign: string = offset < 0 ? '-' : '+';
      this.timezones.push({
        id,
        offset: offset === 0 ? 'UTC' : `UTC ${sign}${formatTwoDigits(hourOffset)}:${formatTwoDigits(minOffset)}`,
      });
    }

    this.timezone = this.timezoneService.guessTimezone();
  }

  /**
   * Auto-detects timezone for instructor.
   */
  onAutoDetectTimezone(): void {
    if (!this.isEnabled) {
      return;
    }
    this.timezone = this.timezoneService.guessTimezone();
  }

  /**
   * Submits the data to add the new course.
   */
  onSubmit(): void {
    if (!this.isEnabled) {
      return;
    }
    if (!this.newCourseId || !this.newCourseName) {
      this.statusMessageService.showErrorToast(
          'Please make sure you have filled in both Course ID and Name before adding the course!');
      return;
    }

    this.isAddingCourse = true;
    this.courseService.createCourse({
      courseName: this.newCourseName,
      timeZone: this.timezone,
      courseId: this.newCourseId,
    }).pipe(finalize(() => this.isAddingCourse = false)).subscribe(() => {
      this.courseAdded.emit();
      this.statusMessageService.showSuccessToast('The course has been added.');
    }, (resp: ErrorMessageOutput) => {
      this.statusMessageService.showErrorToast(resp.error.message);
    });
    this.newCourseId = '';
    this.newCourseName = '';
    this.timezone = this.timezoneService.guessTimezone();
  }

  /**
   * Handles closing of the edit form.
   */
  closeEditFormHandler(): void {
    this.closeCourseFormEvent.emit();
  }
}
