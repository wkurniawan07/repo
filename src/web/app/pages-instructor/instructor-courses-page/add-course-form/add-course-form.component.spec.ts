import { async, ComponentFixture, TestBed } from '@angular/core/testing';

import { CourseService } from '../../../../services/course.service';
import { StatusMessageService } from '../../../../services/status-message.service';
import { TimezoneService } from '../../../../services/timezone.service';
import { AjaxLoadingModule } from '../../../components/ajax-loading/ajax-loading.module';

import { HttpClientTestingModule } from '@angular/common/http/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterTestingModule } from '@angular/router/testing';
import { NgbModule } from '@ng-bootstrap/ng-bootstrap';

import { of } from 'rxjs';
import { AddCourseFormComponent } from './add-course-form.component';

describe('AddCourseFormComponent', () => {
  let component: AddCourseFormComponent;
  let fixture: ComponentFixture<AddCourseFormComponent>;

  const testCourseId: string = 'CS3281';
  const testCourseName: string = 'Valid course';
  const testTimeZone: string = 'UTC';
  const timeZoneOffsets1: Record<string, number> = { GMT: 0 };

  const spyStatusMessageService: any = {
    showErrorToast: jest.fn(),
    showSuccessToast: jest.fn(),
  };
  const timezoneServiceStub: any = {
    getTzOffsets: jest.fn(() => timeZoneOffsets1),
    guessTimezone: jest.fn(() => 'UTC'),
  };
  const spyCourseService: any = {
    createCourse: jest.fn(() => of({})),
  };

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [AddCourseFormComponent],
      imports: [
        HttpClientTestingModule,
        FormsModule,
        ReactiveFormsModule,
        RouterTestingModule,
        NgbModule,
        AjaxLoadingModule,
      ],
      providers: [
        { provide: StatusMessageService, useValue: spyStatusMessageService },
        { provide: CourseService, useValue: spyCourseService },
        { provide: TimezoneService, useValue: timezoneServiceStub },
      ],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(AddCourseFormComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should snap with default fields', () => {
    // Unable to leave timezone as default field, otherwise the field defaults to the
    // timezone the system is on. This will differ from
    // place to place causing the snapshot to constantly be mismatched.
    component.timezone = testTimeZone;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should snap when not enabled', () => {
    component.isEnabled = false;
    component.timezone = testTimeZone;
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('should call showErrorToast when courseId is blank', () => {
    component.newCourseId = '';
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showErrorToast).toHaveBeenCalled();
  });

  it('should hold added course with valid details', () => {
    component.newCourseId = testCourseId;
    component.newCourseName = testCourseName;
    component.onSubmit();
    fixture.detectChanges();
    expect(spyStatusMessageService.showSuccessToast).toHaveBeenCalled();
  });
});
