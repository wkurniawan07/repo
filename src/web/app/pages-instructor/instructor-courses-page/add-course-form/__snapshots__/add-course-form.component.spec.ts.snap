// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCourseFormComponent should snap when not enabled 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isAddingCourse="false"
  isEnabled="false"
  newCourseId=""
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body text-center"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course ID:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-id"
            maxlength="40"
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course Name:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-name"
            maxlength="64"
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label ngb-tooltip-class"
        >
          <span
            ngbtooltip="You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
          >
             Time Zone:
          </span>
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              id="new-time-zone"
            >
              
              <option
                value="GMT"
              >
                GMT (UTC)
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary btn-margin-left"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-md-9 offset-md-3 text-md-left"
        >
          <button
            class="btn btn-success"
            id="btn-save-course"
          >
            
             Add Course 
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;

exports[`AddCourseFormComponent should snap with default fields 1`] = `
<tm-add-course-form
  closeCourseFormEvent={[Function EventEmitter]}
  courseAdded={[Function EventEmitter]}
  courseService={[Function Object]}
  isAddingCourse="false"
  isEnabled={[Function Boolean]}
  newCourseId=""
  newCourseName=""
  statusMessageService={[Function Object]}
  timezone={[Function String]}
  timezoneService={[Function Object]}
  timezones={[Function Array]}
>
  <div
    class="card border-primary fill-plain"
  >
    <div
      class="close-header"
    >
      <button
        aria-label="Close"
        class="close"
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
      </button>
    </div>
    <div
      class="card-body text-center"
    >
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course ID:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-id"
            maxlength="40"
            placeholder="e.g. CS3215-2013Semester1"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label"
        >
          Course Name:
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <input
            class="form-control ng-untouched ng-pristine ng-valid"
            id="new-course-name"
            maxlength="64"
            placeholder="e.g. Software Engineering"
            type="text"
          />
        </div>
      </div>
      <div
        class="form-group row"
      >
        <label
          class="col-md-3 text-md-right col-form-label ngb-tooltip-class"
        >
          <span
            ngbtooltip="You should not
                  need to change this as it is auto-detected based on your device settings. TEAMMATES automatically adjusts
                  to match the current time offset in your area, including clock changes due to daylight saving time."
          >
             Time Zone:
          </span>
        </label>
        <div
          class="col-md-6 text-md-left"
        >
          <div
            class="input-group"
          >
            <select
              class="form-control ng-untouched ng-pristine ng-valid"
              id="new-time-zone"
            >
              
              <option
                value="GMT"
              >
                GMT (UTC)
              </option>
            </select>
            <span
              class="input-group-btn"
              style="margin-left: 5px;"
            >
              <button
                class="btn btn-primary btn-margin-left"
              >
                Auto Detect
              </button>
            </span>
          </div>
        </div>
      </div>
      <div
        class="form-group row"
      >
        <div
          class="col-md-9 offset-md-3 text-md-left"
        >
          <button
            class="btn btn-success"
            id="btn-save-course"
          >
            
             Add Course 
          </button>
        </div>
      </div>
    </div>
  </div>
</tm-add-course-form>
`;
